### PR TITLE
Implement `given/with` syntax

### PR DESCRIPTION
**Pull request by odersky**
This is an elaboration and full implementation of the docs in #7973. Instead of fiddling with the instance keyword (witness or default) we change the parameter syntax by going back to with, which we had before given.

Using given for parameters has several potential problems:

 - Many commenters felt that this was an over-use of given with too many similarities to the current status where implicit is used for everything.
 - Parameters in given instances were particularly burdensome. This was fixed in #7788, but the fix introduces problems on its own. First, it introduces another irregularity. Second, commenters felt the syntax was confusing because it looked too much like function types.

To start with browsing:

https://github.com/dotty-staging/dotty/blob/change-given-with/docs/docs/reference/contextual/motivation-new.md

### CODE REVIEW

**Comment by odersky** [comment]
A note on indentation: Using `with` for context parameters means we cannot possibly use it as well for starting an indented definition block for classes and similar. I had some doubts about this usage of `with` anyway as it was a bit verbose.  So this PR replaces `with` with `:`. For instance, it's now
```scala
object Foo:
  def bar() = 1
```
instead of
```scala
object Foo with
  def bar() = 1
```
This does not mean we go back to the state where a `:` at the end of a line would always imply braces. That extended system is still available only under -Yindent-colons. But without that option,`:` at end of line is only interpreted at the points where a template body or similar starts. Note that this does not introduce an ambiguity with type ascriptions, since no type ascriptions are allowed at these points. https://github.com/dotty-staging/dotty/blob/change-given-with/docs/docs/reference/other-new-features/indentation-new.md explains the details.

I believe `:` works nicely in this new role. For instance,  here's a test that exercises it heavily.

https://github.com/dotty-staging/dotty/blob/change-given-with/tests/pos/reference/delegates.scala

**Comment by arturopala** [comment]
I did read the provided example test several times and each time feel there is something weird in the proposed syntax: 

- some keywords are nouns (extension), some prepositions (given), I see no clear reason for that
- I would expect keyword `as` to set an alias or name, but here, surprise, it sets a resulting type
- Keyword `with` does not sound like a condition to me, using `when` or similar would make it sound better
- Keyword `extension on` does a good job, clearly communicates the intent and reads well
- Similarly extension method syntax is easy to read

I like and use Scala, and would love it to have the most clear and plain syntax as possible, one everyone tech-savy can read and grasp without searching the internet, pls do not compromise on this!

**Comment by soronpo** [comment]
> It's now
> ```scala
> given _
> given T
> ```
> Thta way it's more regular: A given selector is always followed by something: Either
a wildcard or a type.

Regarding the given import syntax, personally I would much rather have:
`import given foo._` to import all *including* the "givens"
`import foo._` to import all *without* "givens"
I'm rather annoyed by the `import foo.{_, given _}` syntax.

File: docs/docs/reference/contextual/givens.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+```
+This creates a given `global` of type `ExecutionContext` that resolves to the right
+hand side `new ForkJoinPool()`.
+The first time `global` is accessed, a new `ForkJoinPool` is created, which is then
+returned for this and all subsequent accesses to `global`.
+
+Alias givens can be anonymous, e.g.
+```scala
+given as Position = enclosingTree.position
+given with outer: Context as Context = outer.withOwner(currentOwner)
```
Comment (lrytz) [inline docs/docs/reference/contextual/givens.md]: is `with x: T` allowed without parentheses, or should it be `given with (outer: Context) as Context = ...`

Comment (odersky) [reply docs/docs/reference/contextual/givens.md]: Good catch. It needs to come in parens.

**Comment by jdegoes** [comment]
I was a proponent of splitting up `given` into two keywords (I suggested `provide` / `given`, e.g. `provide Ord[List[T]] given Ord[T]`), and I think `given` / `with` works about as well as any other choices, so huge 👍 on this from me.

And while subjective, I also think the move away from `with` as the keyword introducing a code block is more aesthetically pleasing, and more inline with other whitespace-sensitive grammars.

The only thing I'm not quite sold on is `as`. In the context of SQL, `as` means to give a name to something, which I believe is the most natural way to use it in English. In C#, it casts to a different type. In FP libraries, it maps to a constant value.

I would suggest the following tweak from this:

```scala
given listOrd[T] with (ord: Ord[T]) as Ord[List[T]]
```

...into this:

```scala
given[T] Ord[List[T]] with (ord: Ord[T]) as listOrd
```

which harmonizes `as` with its meaning in SQL (probably the most common meaning), and biases toward unnamed givens, which I think is an extremely good "best practice". 

In the common case, the above would be written as:

```scala
given[T] Ord[List[T]] with Ord[T]
```

which is quite pleasant. Almost as pleasant as, `provide [T] Ord[List[T]] given Ord[T]`. 😉 

The fact that there are different ways to introduce let bindings is a little unsatisfying. But again could be viewed as a feature if the idea is to bias toward unnamed givens, which I think is the correct pedagogical choice (named givens being reserved for advanced users).

**Comment by smarter** [comment]
> given[T] Ord[List[T]] with (ord: Ord[T]) as listOrd

The problem with a syntax like that is that it looks backward when the result type depends on a parameter:

>  given[T] Ord[List[x.T]] with (x: T) as foo

it also would be backward from the way `with ...` is used in regular defs with implicit parameters.

**Comment by jdegoes** [comment]
@smarter Haskell "solves" that problem by essentially swapping the order, like:

```scala
with[T] (ord: Ord[T]) given Ord[List[T]] as listOrd
```

But it has the same problem as Haskell, which is that to parse out the instance type with your eyes, you have to look past the constraints, which can be arbitrarily long.

OTOH you could use convention to format it like so:

```scala
with[T] (ord: Ord[T]) 
  given Ord[List[T]] as listOrd
```

such that additional constraints don't change where you look for the `given`.

Maybe `as` isn't even needed, you could just provide the `given` a name like the `with`, using type ascription syntax:

```scala
with[T] (ord: Ord[T]) 
  given (listOrd: Ord[List[T]])
```

Then at the risk of bike-shedding too much, change `with` to `given`, and `given` to `implies`:

```scala
given[T] (ord: Ord[T]) 
  implies (listOrd: Ord[List[T]])
```

Or possibly `provide`:

```scala
given[T] (ord: Ord[T]) 
  provide (listOrd: Ord[List[T]])
```

**Comment by odersky** [comment]
@jdegoes Regarding `as`, I answered on the contributors thread, copied here:

I’d read
```scala
given ord as Ord[T] { ... }
```
“given ord as the Ord[Int] instance where …”

If the “ord as” part is missing it’s just “given the Ord[Int] instance where…”.

By that interpretation, the chosen syntax corresponds exactly to spelling it out loud.

**Comment by jdegoes** [comment]
@odersky I think that's teachable, which is the most important concern for me. 👍

**Comment by sideeffffect** [comment]
> this PR replaces `with` with `:`

has `=` been considered? It would be more aligned with other parts of the language, like we don't write `def sum(x, y) with x + y` or `def sum(x, y): x + y`, but `def sum(x, y) = x + y`. For example, a class definition would look like this
```scala
class C(x: Int, y: Int) = A(y) with
  def f = x * y
```

File: docs/docs/reference/contextual/context-parameters.md [hunk 2]
```diff
@@ -69,7 +69,7 @@ maximum(xs).with(descending.with(listOrd.with(intOrd)))
 
 ## Multiple With Clauses
 
-There can be several with clauses in a definition and with clauses can be freely interspersed with normal parameters. Example:
+There can be several with clauses in a definition. Example:
```
Comment (soronpo) [inline docs/docs/reference/contextual/context-parameters.md]: It think it is better to write "There can be several `with` clauses"

File: docs/docs/reference/contextual/context-parameters.md [hunk 3]
```diff
@@ -88,6 +88,17 @@ f(global).with(ctx).with(sym, kind)
 ```
 But `f(global).with(sym, kind)` would give a type error.
 
+With clauses can be freely interspersed with normal parameters, but a normal parameter clause cannot
```
Comment (soronpo) [inline docs/docs/reference/contextual/context-parameters.md]: `With` clauses

File: docs/docs/reference/contextual/context-parameters.md [hunk 4]
```diff
@@ -88,6 +88,17 @@ f(global).with(ctx).with(sym, kind)
...
+directly follow a with parameter clause consisting only of types outside parentheses. So the following is illegal:
```
Comment (soronpo) [inline docs/docs/reference/contextual/context-parameters.md]: `with` parameter

**Comment by sideeffffect** [comment]
I also don't understand why we're keeping "context bounds" around -- they're a relict of a time when declaring implicit/given/with parameters was awkward. But those times will soon be gone and we will be able have things like this just fine:
```scala
def aggregate[A] with Monoid[A] (list: List[A]): A =
    ???
```
And there should be just one way of declaring implicit parameters IMHO.

Also, declaring implicit parameters first, and the ordinary after that has also the benefit that it makes passing methods with implicit parameters easier:
```scala
def f(g: List[Int] => Int): X = ???
given as Monoid[Int] = ???
f(aggregate)
```

before one had to do this
```scala
def aggregate[A](list: List[A])(implicit A: Monoid[A]): A = ???
def f(g: List[Int] => Int): X = ???
implicit val xxx: Monoid[Int] = ???
f(aggregate(_))
```

**Comment by odersky** [comment]
@sideeffffect I believe context bounds are still a lot more concise and easier to read than the alternatives. 

The latest commits allow normal parameters after given parameters, but it's fair to say that this point is still contentious.

**Comment by ri-cisse** [comment]
When defining typeclasses, why do we need to a modifier on the params? We can make `given` keyword required only when passing context params.

```scala

// Defining typeclasses

given [T] (o: Ord[T]) => Ord[List[T]] {}

given [T] Ord[T] => Ord[List[T]] {}
given [T: Ord] => Ord[List[T]] {}

given Ord[Int] {}

given [T, U] (t: Ord[T], u: Ord[U]) => Ord[List(T, U)] {}
given [T, U] Ord[T], Ord[U] => Ord[(T, U)] {}

given [T, U] (t: Ord[T]) => (u: Ord[U]) => Ord[List(T, U)] {}
given [T, U] Ord[T] => Ord[U] => Ord[(T, U)] {}

given ExecutionContext = ExecutionContext.global

// with names

given listOrd[T] (o: Ord[T]) => Ord[List[T]] {}

given listOrd[T] Ord[T] => Ord[List[T]] {}
given listOrd[T: Ord] => Ord[List[T]] {}

given intOrd => Ord[Int] {}

given listOrd[T, U] (t: Ord[T], u: Ord[U]) => Ord[List(T, U)] {}
given listOrd[T, U] Ord[T], Ord[U] => Ord[List(T, U)] {}

given listOrd[T, U] (t: Ord[T]) => (u: Ord[U]) => Ord[List(T, U)] {}
given listOrd[T, U] Ord[T] => Ord[U] => Ord[List(T, U)] {}

given ec => ExecutionContext = ExecutionContext.global


// summoning instances

def max[T](lst: List[T])(given o: Ord[T]): Option[T] = ???
def[T] (lst: List[T]) max(given o: Ord[T]): Option[T] = ???
def[T: Ord] (lst: List[T]) max: Option[T] = ???

extension [T] on (s: List[T])(given Ord[T]) {
  def max: Option[T] = ???
}

extension [T: Ord] on (s: List[T]) {
  def max: Option[T] = ???
}

@main main = {

  max(List(1,2))
  max(List(1,2))(given summon[Ord[Int]])
}

```

**Comment by odersky** [comment]
@ri-cisse When defining typeclasses, why do we need to a modifier on the params? We can make given keyword required only when passing context params.

That is in fact what's currently implemented. People objected on the grounds that it clauses in given instances looked like function types. I am not so sure about this but anyway, using `with` it looks also fine; there's no reason to invent different syntax for conditional given instances.

**Comment by odersky** [comment]
If there is no further input I'd like to merge this by tomorrow. Not sure it's feasible to review this (it's quite a large delta and not very interesting technically) but if someone wants to review parts of it, please go ahead.

**Comment by jdegoes** [comment]
👍 for merging.

File: docs/docs/reference/contextual/motivation-new.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+Existing Scala programmers by and large have gotten used to the status quo and see little need for change. But for newcomers this status quo presents a big hurdle. I believe if we want to overcome that hurdle, we should take a step back and allow ourselves to consider a radically new design.
+
+### The New Design
+
+The following pages introduce a redesign of contextual abstractions in Scala. They introduce four fundamental changes:
+
+ 1. [Given Instances](./givens.md) are a new way to define basic terms that can be synthesized.  They replace implicit definitions. The core principle of the proposal is that, rather than mixing the `implicit` modifier with a large number of features, we have a single way to define terms that can be synthesized for types.
+
+ 2. [With Clauses](./context-parameters.md) define are a new syntax for implicit _parameters_ and their _arguments_. It unambiguously aligns parameters and arguments, solving a number of language warts. It also allows us to have several with clauses in a definition.
```
Comment (aesteve) [inline docs/docs/reference/contextual/motivation-new.md]:
```suggestion
 2. [With Clauses](./context-parameters.md) are a new syntax for implicit _parameters_ and their _arguments_. It unambiguously aligns parameters and arguments, solving a number of language warts. It also allows us to have several with clauses in a definition.
```

Comment (aesteve) [reply docs/docs/reference/contextual/motivation-new.md]: "define a new syntax" also works though, but not both.

**Comment by smarter** [comment]
Now that we have a very lean syntax for implicit parameters, can we drop https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/ShortcutImplicits.scala ? It optimizes methods whose result type is an implicit function type but those can now be written almost as succinctly using `with ...`, and the implementation of ShortcutImplicits is very complex and probably still buggy (I remember someone hitting a crash in it on gitter recently).

**Comment by odersky** [comment]
@smarter lean syntax helps, but if you have many context parameters it's still needless duplication. There's value in abstraction if the abstraction is zero cost. Right now `ShortcutImplicits` duplicates code, so it is not really zero cost. I was thinking that maybe we can use a slightly different approach, where we always use the "shortcut" representation instead of the curried one and rely on bridge construction to keep the right overriding relationships. That would mean rolling ShortcutImplicits into erasure.

**Pull request merged**